File: src/types/metricCourtScene.ts
export type SceneCalibrationStatus = "draft" | "ready" | "degraded" | "invalidated";
export type SceneCalibrationSource = "manual" | "auto_suggested" | "manual_verified";
export type SceneHeightSource = "standard" | "measured";
export type CameraModelSource = "net_refined_virtual" | "homography_constrained_virtual";
export type MetricValidity = "metric_multiview" | "approximate_multiview" | "visualization_only" | "unavailable";

export interface SceneImagePoint {
  x: number;
  y: number;
}

export interface ScenePoint3D {
  x: number;
  y: number;
  z: number;
}

export interface NetProfileControlPoint {
  id: string;
  world: ScenePoint3D;
  image_by_view?: Record<string, SceneImagePoint>;
  provenance?: SceneCalibrationSource;
  confirmed?: boolean;
}

export interface NetProfile {
  profile_type: "standard" | "measured";
  height_source: SceneHeightSource;
  coordinate_units: "feet";
  control_points: NetProfileControlPoint[];
  sampled_top_profile: ScenePoint3D[];
  post_world_points?: ScenePoint3D[];
}

export interface SceneCameraQuality {
  court_reprojection_error_px?: number | null;
  net_reprojection_error_px?: number | null;
  holdout_reprojection_error_px?: number | null;
  ray_angle_deg?: number | null;
  height_uncertainty_ft?: number | null;
  status: "ok" | "warning" | "failed";
  rejection_reasons: string[];
}

export interface SceneCameraModel {
  source: CameraModelSource;
  projection?: number[][] | null;
  rotation?: number[][] | null;
  translation?: number[] | null;
  focal_px?: number | null;
  image_width?: number | null;
  image_height?: number | null;
}

export interface SceneViewCalibration {
  view_id: string;
  camera_id?: string | null;
  video_id?: string | null;
  calibration_id?: string | null;
  image_width?: number | null;
  image_height?: number | null;
  frame_index?: number | null;
  timestamp_seconds?: number | null;
  court_orientation?: string | null;
  net_annotations: Record<string, SceneImagePoint>;
  holdout_annotations?: Record<string, SceneImagePoint>;
  camera_model?: SceneCameraModel | null;
  quality: SceneCameraQuality;
  provenance: SceneCalibrationSource;
}

export interface MetricCourtSceneCalibration {
  schema_version: "metric_court_scene.v1";
  capture_take_id: string;
  revision: number;
  status: SceneCalibrationStatus;
  canonical_frame_id?: string | null;
  coordinate_units: "feet";
  court_width_ft: number;
  court_length_ft: number;
  net_y_ft: number;
  net_profile: NetProfile;
  holdout_control_points?: NetProfileControlPoint[];
  views: SceneViewCalibration[];
  provenance: SceneCalibrationSource;
  quality: SceneCameraQuality;
  fallback_metric_validity: MetricValidity;
  created_at: string;
  updated_at: string;
  published_at?: string | null;
  rejection_reasons: string[];
}

export interface MetricCourtSceneDraftRequest {
  canonical_frame_id?: string | null;
  net_profile?: Partial<NetProfile> & Pick<NetProfile, "control_points">;
  holdout_control_points?: NetProfileControlPoint[];
  views: SceneViewCalibration[];
  provenance?: SceneCalibrationSource;
}

export interface MetricCourtSceneValidationResponse {
  capture_take_id: string;
  status: Exclude<SceneCalibrationStatus, "draft">;
  quality: SceneCameraQuality;
  rejection_reasons: string[];
  scene: MetricCourtSceneCalibration;
}

export interface MetricCourtSceneRevisionSummary {
  revision: number;
  status: SceneCalibrationStatus;
  provenance: SceneCalibrationSource;
  created_at: string;
  published_at?: string | null;
}

export const STANDARD_NET_HEIGHT_FT = {
  endpoint: 36 / 12,
  center: 34 / 12,
} as const;

export function buildStandardNetProfile(confirmed = false): NetProfile {
  const control_points: NetProfileControlPoint[] = [
    { id: "left", world: { x: 0, y: 22, z: STANDARD_NET_HEIGHT_FT.endpoint }, provenance: "manual", confirmed },
    { id: "center", world: { x: 10, y: 22, z: STANDARD_NET_HEIGHT_FT.center }, provenance: "manual", confirmed },
    { id: "right", world: { x: 20, y: 22, z: STANDARD_NET_HEIGHT_FT.endpoint }, provenance: "manual", confirmed },
  ];
  return {
    profile_type: "standard",
    height_source: "standard",
    coordinate_units: "feet",
    control_points,
    sampled_top_profile: [],
    post_world_points: [
      { x: -1, y: 22, z: STANDARD_NET_HEIGHT_FT.endpoint },
      { x: 21, y: 22, z: STANDARD_NET_HEIGHT_FT.endpoint },
    ],
  };
}

File: backend/app/schemas/metric_court_scene.py
"""Metric court scene calibration contracts.

The first version is intentionally file-backed and capture-take scoped.  It
keeps the static court/net geometry separate from AnalysisJob artifacts so a
fixed camera setup can be reused by multiple recordings and re-runs.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


SceneCalibrationStatus = Literal["draft", "ready", "degraded", "invalidated"]
SceneCalibrationSource = Literal["manual", "auto_suggested", "manual_verified"]
SceneHeightSource = Literal["standard", "measured"]
CameraModelSource = Literal["net_refined_virtual", "homography_constrained_virtual"]
MetricValidity = Literal["metric_multiview", "approximate_multiview", "visualization_only", "unavailable"]


class SceneImagePoint(BaseModel):
    x: float
    y: float


class ScenePoint3D(BaseModel):
    x: float
    y: float
    z: float


class NetProfileControlPoint(BaseModel):
    id: str
    world: ScenePoint3D
    image_by_view: dict[str, SceneImagePoint] = Field(default_factory=dict)
    provenance: SceneCalibrationSource = "manual"
    confirmed: bool = False


class NetProfile(BaseModel):
    profile_type: Literal["standard", "measured"] = "standard"
    height_source: SceneHeightSource = "standard"
    coordinate_units: Literal["feet"] = "feet"
    control_points: list[NetProfileControlPoint] = Field(default_factory=list)
    sampled_top_profile: list[ScenePoint3D] = Field(default_factory=list)
    post_world_points: list[ScenePoint3D] = Field(default_factory=list)


class SceneCameraQuality(BaseModel):
    court_reprojection_error_px: float | None = Field(default=None, ge=0.0)
    net_reprojection_error_px: float | None = Field(default=None, ge=0.0)
    holdout_reprojection_error_px: float | None = Field(default=None, ge=0.0)
    ray_angle_deg: float | None = Field(default=None, ge=0.0)
    height_uncertainty_ft: float | None = Field(default=None, ge=0.0)
    status: Literal["ok", "warning", "failed"] = "warning"
    rejection_reasons: list[str] = Field(default_factory=list)


class SceneCameraModel(BaseModel):
    source: CameraModelSource = "homography_constrained_virtual"
    projection: list[list[float]] | None = None
    rotation: list[list[float]] | None = None
    translation: list[float] | None = None
    focal_px: float | None = Field(default=None, ge=0.0)
    image_width: int | None = Field(default=None, ge=0)
    image_height: int | None = Field(default=None, ge=0)


class SceneViewCalibration(BaseModel):
    view_id: str
    camera_id: str | None = None
    video_id: str | None = None
    calibration_id: str | None = None
    image_width: int | None = Field(default=None, ge=0)
    image_height: int | None = Field(default=None, ge=0)
    frame_index: int | None = Field(default=None, ge=0)
    timestamp_seconds: float | None = Field(default=None, ge=0.0)
    court_orientation: str | None = None
    net_annotations: dict[str, SceneImagePoint] = Field(default_factory=dict)
    holdout_annotations: dict[str, SceneImagePoint] = Field(default_factory=dict)
    camera_model: SceneCameraModel | None = None
    quality: SceneCameraQuality = Field(default_factory=SceneCameraQuality)
    provenance: SceneCalibrationSource = "manual"


class MetricCourtSceneCalibration(BaseModel):
    schema_version: Literal["metric_court_scene.v1"] = "metric_court_scene.v1"
    capture_take_id: str
    revision: int = Field(default=0, ge=0)
    status: SceneCalibrationStatus = "draft"
    canonical_frame_id: str | None = None
    coordinate_units: Literal["feet"] = "feet"
    court_width_ft: float = Field(default=20.0, gt=0.0)
    court_length_ft: float = Field(default=44.0, gt=0.0)
    net_y_ft: float = Field(default=22.0, ge=0.0, le=44.0)
    net_profile: NetProfile = Field(default_factory=NetProfile)
    holdout_control_points: list[NetProfileControlPoint] = Field(default_factory=list)
    views: list[SceneViewCalibration] = Field(default_factory=list)
    provenance: SceneCalibrationSource = "manual"
    quality: SceneCameraQuality = Field(default_factory=SceneCameraQuality)
    fallback_metric_validity: MetricValidity = "unavailable"
    created_at: datetime
    updated_at: datetime
    published_at: datetime | None = None
    rejection_reasons: list[str] = Field(default_factory=list)


class MetricCourtSceneDraftRequest(BaseModel):
    canonical_frame_id: str | None = None
    net_profile: NetProfile = Field(default_factory=NetProfile)
    holdout_control_points: list[NetProfileControlPoint] = Field(default_factory=list)
    views: list[SceneViewCalibration] = Field(default_factory=list)
    provenance: SceneCalibrationSource = "manual"


class MetricCourtSceneValidationResponse(BaseModel):
    capture_take_id: str
    status: Literal["ready", "degraded", "invalidated"]
    quality: SceneCameraQuality
    rejection_reasons: list[str] = Field(default_factory=list)
    scene: MetricCourtSceneCalibration


class MetricCourtSceneRevisionSummary(BaseModel):
    revision: int
    status: SceneCalibrationStatus
    provenance: SceneCalibrationSource
    created_at: datetime
    published_at: datetime | None = None
